Documentation/anti-patterns.md

```diff
@@ -0,0 +1,180 @@
+In fact, `Dial` does not always establish a connection to servers by default. The connection
+behavior is determined by the load balancing policy being used. For instance, an "active"
+load balancing policy such as Round Robin attempts to maintain a constant connection, while the default "pick first"
+policy delays connection until an RPC is executed. To explicitly initiate a connection, you can employ
+the WithBlock option.
+
+### Using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+The gRPC API provides several options that can be used to configure the behavior
+of dialing and connecting to a gRPC server. Some of these options, such as
+`FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`, rely on
+failures at dial time. However, we strongly discourage developers from using
+these options, as they can introduce race conditions and result in unreliable
+and difficult-to-debug code.
+
+###  Why we discourage using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+When a client attempts to connect to a gRPC server, it can encounter a variety
+of errors, including network connectivity issues, server-side errors, and
+incorrect usage of the gRPC API. The options `FailOnNonTempDialError`,
+`WithBlock`, and `WithReturnConnectionError` are designed to handle some of
+these errors, but they do so by relying on failures at dial time. This means
+that they may not provide reliable or accurate information about the status of
+the connection.
+
+For example, if a client uses `WithBlock` to wait for a connection to be
+established, it may end up waiting indefinitely if the server is not responding.
+Similarly, if a client uses `WithReturnConnectionError` to return a connection
+error if dialing fails, it may miss opportunities to recover from transient
+network issues that are resolved shortly after the initial dial attempt.
+
+## Best practices for error handling in gRPC
+
+Instead of relying on failures at dial time, we strongly encourage developers to
+rely on errors from RPCs. When a client makes an RPC, it can receive an error
+response from the server. These errors can provide valuable information about
+what went wrong, including information about network issues, server-side errors,
+and incorrect usage of the gRPC API.
+
+By handling errors from RPCs correctly, developers can write more reliable and
+robust gRPC applications. Here are some best practices for error handling in
+gRPC:
+
+- Always check for error responses from RPCs and handle them appropriately.
+- Use the `status` field of the error response to determine the type of error that
+  occurred.
+- When retrying failed RPCs, use a backoff strategy to avoid
+  overwhelming the server or exacerbating network issues.
+- Avoid using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`,
+  as these options can introduce race conditions and result in unreliable and
+  difficult-to-debug code.
+
```

Comment:
Add something like this?

- If making the outgoing RPC in order to handle an incoming RPC, be sure to translate the status code before returning the error from your method handler.  For example, if the error is an INVALID_ARGUMENT error, that probably means your service has a bug (otherwise it shouldn't have triggered this error), in which case INTERNAL is more appropriate to return back to your users.